v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,81 @@
+
+#include <stdio.h>
+#include <stdarg.h>
+
+#include "windows.h"
+
+#include "TraceLoggingProvider.h"
+#include "evntrace.h"
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/logger.h"
+#include "c_logging/log_sink_etw.h"
+
+TRACELOGGING_DEFINE_PROVIDER(
+    g_my_component_provider,
+    "block_storage_2",
+    (0xDAD29F36, 0x0A48, 0x4DEF, 0x9D, 0x50, 0x8E, 0xF9, 0x03, 0x6B, 0x92, 0xB4));
+/*DAD29F36-0A48-4DEF-9D50-8EF9036B92B4*/
+
+static volatile LONG isETWLoggerRegistered = 0;
+
+static void lazyRegisterEventProvider(void)
+{
+    /*lazily init the logger*/
+    LONG state;
+    while ((state = InterlockedCompareExchange(&isETWLoggerRegistered, 1, 0)) != 2) /*0 - not init, 1 - initializing, 2 - initialized*/
+    {
+        if (state == 0)
+        {
+            /* register the provider*/
+            TLG_STATUS t = TraceLoggingRegister(g_my_component_provider);
+            if (SUCCEEDED(t))
+            {
+                (void)InterlockedExchange(&isETWLoggerRegistered, 2);
+                printf("block_storage_2 ETW provider was registered succesfully (self test). Executable file full path name = %s", _pgmptr); /*_pgmptr comes from https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea */
+            }
+            else
+            {
+                (void)printf("block_storage_2 ETW provider was NOT registered.");
+                (void)InterlockedExchange(&isETWLoggerRegistered, 0);
+            }
+        }
+    }
+}
+
+#define ETW_TRACE_LOGGING_WRAPPER(event_name, trace_level, message, file, func, line) \
+    TraceLoggingWrite(g_my_component_provider, \
+        event_name, \
+        TraceLoggingLevel(trace_level), \
+        TraceLoggingString(message, "content"), \
+        TraceLoggingString(file, "file"), \
+        TraceLoggingString(func, "func"), \
+        TraceLoggingInt32(line, "line")) \
+
+void log_sink_etw_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* message, const char* file, const char* func, int line)
+{
+    (void)log_context;
+
+    lazyRegisterEventProvider();
+
+    switch (log_level)
+    {
+    default:
+        ETW_TRACE_LOGGING_WRAPPER("Unknown", TRACE_LEVEL_NONE, message, file, func, line);
+    case LOG_LEVEL_CRITICAL:
+        ETW_TRACE_LOGGING_WRAPPER("LogCritical", TRACE_LEVEL_CRITICAL, message, file, func, line);
+    case LOG_LEVEL_ERROR:
+        ETW_TRACE_LOGGING_WRAPPER("LogError", TRACE_LEVEL_ERROR, message, file, func, line);
+    case LOG_LEVEL_WARNING:
+        ETW_TRACE_LOGGING_WRAPPER("LogWarning", TRACE_LEVEL_WARNING, message, file, func, line);
+    case LOG_LEVEL_INFO:
+        ETW_TRACE_LOGGING_WRAPPER("LogInfo", TRACE_LEVEL_INFORMATION, message, file, func, line);
+    case LOG_LEVEL_VERBOSE:
+        ETW_TRACE_LOGGING_WRAPPER("LogVerbose", TRACE_LEVEL_VERBOSE, message, file, func, line);
+    }
+}
+
+const LOG_SINK etw_log_sink = { log_sink_etw_log };
```

Comment:
>log_sink_etw_log [](http://example.com/codeflow?start=32&length=16)

will wait, here's reaons:
1) while this is an improvement over previous existing logging, i believe we should really think for 1 more day about having the context data:
a) structured
b) maybe evaluated every time when logged (so not only once at context creation time) - if at all possible
2) missing working examples with context being passed on the stack (and can a local context be passed as argument?)
3) try to avoid using malloc at all. Maybe have a define that says that no more than 10 contexts can be nested, and every "new" context is allocated on the stack and copies parent's chain of context.

... I am not sure I am helpful, but these were all my thoughts.

Reply:
aaah! and of course I am curious to see where this is going :)

Reply:
And one more thing: since I cannot add comments in the "descriotion.txt". Can you format those log/s in a more human readable format? I mean, if it does 600K/second, nobody cares about the 5th digit after the decimal point :)

Reply:
1) a) I went back and forth about this a couple of times. I was thinking at some point to do a full blown type system like for ZRPC or umock. I think it would likely be overshoot.
The simplest thing is stringifying everything. Of course that is crappy from the point of view that not even the property-value pairs are separated.
So what I am thinking now we should at least separate the property-value pairs behind and make them all string. That would give us the ability to push the data somewhat more structured to Jarvis and get better queries.
Going full blown type system would likely not be that helpful because also ETW has a limited set of types and we'd end up anyway stringifying many things.

b) Let me tinker with this for a little bit. I am thinking if we would capture addresses for the values, that might also solve the desired deferred copy of memory due to stringifying.

Reply:
2) I'll add that, should not be hard.

3) That's what the local context is for avoiding the malloc. I think chaining local contexts should be doable, let me get a working example of that.

Reply:
Very helpful btw. This is how most of our reviews should go. I feel like we lost some of the RFC part of our process that we used to have. Pride is a big enemy of writing SW.

Reply:
Also updated the perf test to not spew any digits. Btw the perf test will likely not get checked in in its current form, this was simply for me to prove to myself where it is and guide the development.